lib/firenest/replicated_state.ex

```diff
@@ -0,0 +1,624 @@
+  some amount of recent deltas are retained for "catching up" remote nodes.
+  If remote node is too far behind the current state and a delta-based
+  catch-up can't be performed, the server falls back to transferring the entire
+  local state. If keeping track of incremental changes is not convenient
+  for a particular state type, the value of a delta can be set to be equal
+  to the current state - this will always cause full state transfers.
+  """
+  alias Firenest.SyncedServer
+
+  @type server() :: atom()
+  @type key() :: term()
+
+  @type local_delta() :: term()
+  @type remote_delta() :: term()
+  @type state() :: term()
+  @type config() :: term()
+
+  @type delayed_update() :: {timeout :: pos_integer(), update :: term()}
+
+  @doc """
+  Called when a partition starts up.
+
+  It returns the initial delta value used for incremental state change
+  tracking and  an immutable `config` value that will be passed to all
+  callbacks.
+  """
+  @callback init(opts :: keyword()) :: {initial_delta :: local_delta(), config()}
+
+  @doc """
+  Called whenever the `put/4` function is called to create a new state.
+
+  The `arg` is received from the corresponding `put/4` call. For the
+  explanation of the `delayed_update` and `:delete` return values see
+  the `c:local_update/4` callback.
+
+  This is a good place for broadcasting local state changes.
+  """
+  @callback local_put(arg :: term(), config()) ::
+              {:ok, initial_state} | {:ok, initial_state, delayed_update() | :delete}
+            when initial_state: state
+
+  @doc """
+  Called whenever the `update/4` function is called to update a state.
+
+  This is a good place for broadcasting local state changes.
+
+  ## Delayed update and delete
+
+  If the function returns a third element in the tuple consisting of
+  `{timeout, update}`, a timer is started by the server and the
+  `c:local_update/4` callback will be called with the `update` value
+  after `timeout` milliseconds.
+
+  If the third element is `:delete`, the state will be immediately
+  deleted and the `c:local_delete/2` callback triggered.
+  """
+  @callback local_update(update :: term(), local_delta(), state(), config()) ::
+              {local_delta(), state()} | {local_delta(), state(), delayed_update() | :delete}
+
+  @doc """
+  Called whenever attached process dies or the `delete/3` or `delete/2`
+  function is called and state is about to be deleted.
+
+  The return value is ignored.
+
+  This is a good place for broadcasting local state changes.
+  """
+  @callback local_delete(state(), config()) :: term()
+
+  @doc """
+  Called whenever the server is about to incrementally replicate local
+  state to a remote node.
+
+  It takes in the local delta value constructed in the `c:local_update/4`
+  calls and returns a remote delta value that will be replicated to other
+  servers. The value of the local delta is reset to the initial delta value
```

Comment:
I'm pretty sure I'm missing something obvious here 😅 Does it mean that `local_delta` argument will be the delta returned by `local_update/4`?